map/client/client.go

```diff
@@ -0,0 +1,335 @@
+type Client struct {
+	Session *client.Session
+}
+
+func NewClient(s *client.Session) (*Client, error) {
+	return &Client{Session: s}, nil
+}
+
+type StorageLocation interface {
+	ID() common.MessageID
+	Name() string
+	Provider() string
+}
+
+type mapStorage struct {
+	id             common.MessageID
+	name, provider string
+}
+
+func (m *mapStorage) ID() common.MessageID {
+	return m.id
+}
+func (m *mapStorage) Name() string {
+	return m.name
+}
+func (m *mapStorage) Provider() string {
+	return m.provider
+}
+
+type DeterministicDescriptorList []utils.ServiceDescriptor
+
+func (d DeterministicDescriptorList) Less(i, j int) bool {
+	return d[i].Name+d[i].Provider < d[j].Name+d[j].Provider
+}
+func (d DeterministicDescriptorList) Len() int {
+	return len(d)
+}
+func (d DeterministicDescriptorList) Swap(i, j int) {
+	d[i], d[j] = d[j], d[i]
+}
+
+func deterministicSelect(descs []utils.ServiceDescriptor, slot int) utils.ServiceDescriptor {
+	// TODO: order the descriptors by their identity keys
+	ddescs := DeterministicDescriptorList(descs)
+	sort.Sort(ddescs)
+	return ddescs[slot]
+}
+
+// GetStorageProvider returns the deterministically selected storage provider given a storage secret ID
+func (c *Client) GetStorageProvider(ID common.MessageID) (StorageLocation, error) {
+	// doc must be current document!
+	doc := c.Session.CurrentDocument()
+	if doc == nil {
+		return nil, errors.New("No PKI document") // XXX: find correct error
+	}
+	descs := utils.FindServices(common.MapServiceName, doc)
+	if len(descs) == 0 {
+		return nil, errors.New("No descriptors")
+	}
+	slot := int(binary.LittleEndian.Uint64(ID[:8])) % len(descs)
+	// sort the descs and return the chosen one
+	desc := deterministicSelect(descs, slot)
+	return &mapStorage{name: desc.Name, provider: desc.Provider, id: ID}, nil
+}
+
+// Put places a value into the store
+func (c *Client) Put(ID common.MessageID, signature, payload []byte) error {
+	if !ID.WritePk().Verify(signature, payload) {
+		return errors.New("signature does not verify Write")
+	}
+
+	loc, err := c.GetStorageProvider(ID)
+	if err != nil {
+		return err
+	}
+	b := common.MapRequest{ID: ID, Signature: signature, Payload: payload}
+	// XXX: ideally we limit the number of retries
+	// so that it doesn't keep trying to deliver to a stale/missing service forever...
+	serialized, err := cbor.Marshal(b)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.Session.SendUnreliableMessage(loc.Name(), loc.Provider(), serialized)
+	// XXX: do we need to track msgId and see if it was delivered or not ???
+	return err
+}
+
+// PayloadSize returns the size of the user payload
+func (c *Client) PayloadSize() int {
+	geo := c.Session.SphinxGeometry()
+	return geo.UserForwardPayloadLength - cborFrameOverhead - eddsa.SignatureSize
+}
+
+// Get requests ID from the chosen storage node and returns a payload or error
+func (c *Client) Get(ID common.MessageID, signature []byte) ([]byte, error) {
+
+	if !ID.ReadPk().Verify(signature, ID[:]) {
```

Comment:
here is another example of mixing IO with cryptography which would in some contexts be totally fine however in this case that's not a cryptography operation that is woven into the core cryptography module.